receiver/telemetrygeneratorreceiver/generators.go

```diff
@@ -0,0 +1,69 @@
+// Copyright observIQ, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package telemetrygeneratorreceiver //import "github.com/observiq/bindplane-agent/receiver/telemetrygeneratorreceiver"
+
+import (
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+)
+
+// GeneratorType is the type of generator to use, either "logs", "host_metrics", or "windows_events"
+type GeneratorType = string
+
+const (
+	// GeneratorTypeLogs is the generator type for logs
+	GeneratorTypeLogs GeneratorType = "logs"
+
+	// GeneratorTypeHostMetrics is the generator type for host metrics
+	GeneratorTypeHostMetrics GeneratorType = "host_metrics"
+
+	// GeneratorTypeWindowsEvents is the generator type for windows events
+	GeneratorTypeWindowsEvents GeneratorType = "windows_events"
+)
+
+type generator interface {
+	// SupportsType returns true if the generator supports the given component type, either metrics, logs, or traces.
+	SupportsType(component.Type) bool
+	GenerateMetrics() pmetric.Metrics
+	GenerateLogs() plog.Logs
+	GenerateTraces() ptrace.Traces
```

Comment:
@BinaryFissionGames Curious if you have a proposed design. I went with this because of the types returned. I tried to think of a generic Generate function.

Are you purposing something like this but one for each type?

```go
type metricsGenerator interface {
    Generate() pmetric.Metrics
}
```

Reply:
Yeah, I thought about this a little more and I was thinking it would make more sense to have 3 interfaces, one for each type

```go
type metricsGenerator interface {
    GenerateMetrics() pmetric.Metrics
}
```
Then one for logs and one for traces.

Then you can just have a 
`map[string]metricsGeneratorCreateFunc` 
(`type metricsGeneratorCreateFunc func() metricsGenerator`), which would make it easy to:
1. check if a certain generator exists for the given telemetry type
2. create the generator from the type in a typesafe manner

Reply:
This is how I'd see it being setup upstream, it's similar to how something like consumers are set up in OTel, like you have ConsumeMetrics, ConsumeLogs, etc.

Reply:
How does upstream do it? Do they use reflection to see if an object implements an interface? I'm thinking if I have say two interfaces and configured all three pipeline types do I have to check at runtime to see if a struct implements an interface before trying to pass it in as such?

Reply:
So, I think I missed that we were trying to re-use the type for each pipeline when I suggested this.

I think they upstream DOES avoid reflection, but I can't find where they reuse already-created components, if they do (the pipeline creation code is kinda complicated, so I easily could have just missed it).

Personally, I'm in favor of the type assertions here, I feel like in principle it's not really different from the SupportsType function, and if we're gonna be creating a lot of generators it'd be nice to not have to create stub methods for telemetry types that are not used (e.g. I create a logs-only generator and now have to stub out traces and metrics).

That being said, I get that their are downside to that approach, so if we feel like the cons of the type assertion is bad and we feel like we need to use the same struct across traces, metrics, and logs, then I'm OK with the current approach.

Reply:
Yeah, I agree with the fact of having extra stub methods. It makes the interface feel too robust. @shazlehu Can we switch to the design Brandon has suggested?

@BinaryFissionGames Also, I wonder if they just do a type check. Can you cast to an interface an interface as a way of seeing if you implement it?

Reply:
You're talking about a type assertion like `mGen, ok := generator.(metricsGenerator)`, right? That's what I mean to suggest. It's how stuff like e.g. error unwrapping is handled in the go stdlib 

https://cs.opensource.google/go/go/+/refs/tags/go1.22.0:src/errors/wrap.go;l=17

Reply:
I'd actually like if we verified somehow that when this is included in multiple pipeline types, that it doesn't create separate receivers. I think it actually does, and doing this type assertion thing is maybe actually pointless.

Reply:
>I'd actually like if we verified somehow that when this is included in multiple pipeline types, that it doesn't create separate receivers.

Yeah, I forgot about this. It does create a new receiver per pipeline. Generators don't know what pipeline their in so we need some sort of type assertion or check to see if a generator is compatible with the pipeline it's configured for.